viofs/pci/power.c

```diff
@@ -0,0 +1,228 @@
+ * Written By: Gal Hammer <ghammer@redhat.com>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met :
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and / or other materials provided with the distribution.
+ * 3. Neither the names of the copyright holders nor the names of their contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "viofs.h"
+
+#include "power.tmh"
+
+#ifdef ALLOC_PRAGMA
+#pragma alloc_text(PAGE, VirtFsEvtDevicePrepareHardware)
+#pragma alloc_text(PAGE, VirtFsEvtDeviceReleaseHardware)
+#pragma alloc_text(PAGE, VirtFsEvtDeviceD0Entry)
+#pragma alloc_text(PAGE, VirtFsEvtDeviceD0Exit)
+#endif
+
+NTSTATUS VirtFsEvtDevicePrepareHardware(IN WDFDEVICE Device,
+                                        IN WDFCMRESLIST Resources,
+                                        IN WDFCMRESLIST ResourcesTranslated)
+{
+    PDEVICE_CONTEXT context = GetDeviceContext(Device);
+    NTSTATUS status = STATUS_SUCCESS;
+	u64 HostFeatures, GuestFeatures = 0;
+
+    UNREFERENCED_PARAMETER(Resources);
+
+    TraceEvents(TRACE_LEVEL_VERBOSE, DBG_POWER, "--> %!FUNC! Device: %p",
+        Device);
+
+    PAGED_CODE();
+
+    status = VirtIOWdfInitialize(&context->VDevice, Device,
+		ResourcesTranslated, NULL, VIRT_FS_MEMORY_TAG);
+
+	if (!NT_SUCCESS(status))
+    {
+        TraceEvents(TRACE_LEVEL_ERROR, DBG_POWER,
+			"VirtIOWdfInitialize failed with %!STATUS!", status);
+    }
+
+	HostFeatures = VirtIOWdfGetDeviceFeatures(&context->VDevice);
+
+	if (virtio_is_feature_enabled(HostFeatures, VIRTIO_F_VERSION_1))
+	{
+		virtio_feature_enable(GuestFeatures, VIRTIO_F_VERSION_1);
+	}
+	if (virtio_is_feature_enabled(HostFeatures, VIRTIO_F_ANY_LAYOUT))
+	{
+		virtio_feature_enable(GuestFeatures, VIRTIO_F_ANY_LAYOUT);
+	}
+
```

Comment:
The driver does not need to set explicitly VIRTIO_F_VERSION_1 and VIRTIO_F_ANY_LAYOUT, the WDF library does this when needed.